### ISSUE TITLE
Compatability with aws-sdk-php v3.2.1

### ISSUE THREAD

**Issue by ixalon** [deliciousbrains/wp-amazon-s3-and-cloudfront#267]
We're using aws-sdk-php v3.2.1 in our own in-house plugins (bought in using composer). Unfortunately using this version appears to clash with amazon-s3-and-cloudfront. With our plugin(s) enabled, we see the following error when trying to list or create buckets:

`Error fetching buckets — Missing required client configuration options: region: (string) A "region" configuration value is required for the "s3" service (e.g., "us-west-2"). A list of available public regions and endpoints can be found at http://docs.aws.amazon.com/general/latest/gr/rande.html. version: (string) A "version" configuration value is required. Specifying a version constraint ensures that your code will not be affected by a breaking change made to the service. For example, when using Amazon S3, you can lock your API version to "2006-03-01". Your build of the SDK has the following version(s) of "s3": * "2006-03-01" You may provide "latest" to the "version" configuration value to utilize the most recent available API version that your client's API provider can find. Note: Using 'latest' in a production application is not recommended. A list of available API versions can be found on each client's API documentation page: http://docs.aws.amazon.com/aws-sdk-php/v3/api/index.html. If you are unable to load a specific API version, then you may need to update your copy of the SDK.`

Specifying AWS_REGION in wp-config.php does not help (despite what the amazon-web-services' plugin's readme says - again, might be a SDK version issue).

Patching `get_s3client()` to always pass a region and API version fixes the issue.

**Comment by toneby** [comment]
I have the same problem with aws-sdk-php version 3.2.6. I added the changes Ixalon already have added but for some reason it was not enough, and I don't have time to debug :(. But my temporary fix was to hard code the region in lots of places. Not a future proof solution, but hopefully it will work as expected in the next release.

**Comment by A5hleyRich** [comment]
How are you including the AWS SDK for use with WP Offload S3? Plus, is there any specific features you're using from the 3.* release?

**Comment by toneby** [comment]
I'm simply using it to fetch a tag so nothing special at all, so it would be possible to downgrade to whatever you recommend I suppose, although I would prefer not to have to fiddle with it more than necessary.

I include it using composer.json and let AWS take care of the installation on the EB instances. My composer looks like this:
```
{
    "require": {
        "aws/aws-sdk-php": "3.2.6"
    }
}
```

I updated to the latest version (0.9.3) just now so I'll have a closer look at where I should change the code and make myself a diff until you release a fix for the problems.

**Comment by toneby** [comment]
I've just downgraded to 2.8.3 (that seems to be what you are using) and I'm still seeing the same error message. So somehow, for some reason Elastic Beanstalk manages to get v3 of the SDK. I have no idea as to how to solve this. I've also made sure that the versions of amazon-s3-and-cloudfront and amazon-web-services are the latest and unchanged by me.

/Peter

**Comment by ixalon** [comment]
For me, v3 of the SDK is being bought into the php runtime by a wordpress plugin we've built ourselves. It too has a composer.json, containing:
```
{
    "require": {
        "aws/aws-sdk-php": "3.2.1"
    }
}
```
Our CI environment runs composer and builds a .zip of the plugin for internal distribution containing all our code, autoload.php and all dependencies.

If Wordpress includes our plugin before yours then the PHP classes get defined from the V3 installation in our plugin. If it includes your plugin first, then V2 is used.

However we could change to use v2.8.6 (we rely on a change in this version).

**Comment by toneby** [comment]
I just found the reason for my problems. The configuration stored in the database by WP contained the above error as the value for "region". 
To check for that do the following:
Connect to you database and run the following query:
`select * from wpp_options where option_name='tantan_wordpress_s3'`
if the string returned contains the above error simply do a:
`delete from wpp_options where option_name='tantan_wordpress_s3'`
You should probably make a backup of the row before deleting it, just in case.

**Comment by A5hleyRich** [comment]
Planned for next major release of WP Offload S3.

**Comment by bashquandry** [comment]
I've been getting this same error from above for the last few months on all the sites I use WP Offload. It's not my policy because I've followed the instructions to the letter and also just flat out set it to Admin Level and Full Access. I have yet to find a solution posted. The error is as follows:

Error creating bucket — Missing required client configuration options: region: (string) A "region" configuration value is required for the "s3" service (e.g., "us-west-2"). A list of available public regions and endpoints can be found at http://docs.aws.amazon.com/general/latest/gr/rande.html. version: (string) A "version" configuration value is required. Specifying a version constraint ensures that your code will not be affected by a breaking change made to the service. For example, when using Amazon S3, you can lock your API version to "2006-03-01". Your build of the SDK has the following version(s) of "s3": * "2006-03-01" You may provide "latest" to the "version" configuration value to utilize the most recent available API version that your client's API provider can find. Note: Using 'latest' in a production application is not recommended. A list of available API versions can be found on each client's API documentation page: http://docs.aws.amazon.com/aws-sdk-php/v3/api/index.html. If you are unable to load a specific API version, then you may need to update your copy of the SDK.

**Comment by kevinsperrine** [comment]
+1

**Comment by divyatechryde** [comment]
Hi , please suggest solution to this .

I have installed AWS as plugin and wp-amazon-s3-and-cloudfront , also deleted and reinstalled to get the latest updates.

Do i need to install AWS php sdk separately if it is not working on latest version. 

Please suggest the workaround. 

Also deleting option_name from wordpress table did not help as that value gets recreated , though i edit the php file under classes however no joy.

**Comment by A5hleyRich** [comment]
@divyatechryde AWS SDK v3 hasn't been implemented yet.

**Comment by chredd** [comment]
Hello! 
+1 on this feature. It is causing a lot of problems in a project that I am currently working on.

Thanks for a great plugin!

**Comment by elgeorge2k** [comment]
I just faced the same problem! just realized it had this error AFTER DEPLOY TO LIVE!

**Comment by andrewmclagan** [comment]
big plus one on this issue

**Comment by mikecowgill** [comment]
I found a filter in the amazon-web-services.php class called **aws_get_client_args** that allows us to override the aws client args. These args should work for aws sdk v3:

function set_aws_client_args( $args ) {

	// overwrite the args. With aws api 3+ the credentials are stored differently as well
	$args = array(
        'credentials' => array(
            'key' => '*****************',
            'secret'  => '**************************',
        ),
        'region'  => 'us-west-2',
        'version' => 'latest',
    );
	
    return $args;

}
add_filter('aws_get_client_args', 'set_aws_client_args');

**Comment by aaemnnosttv** [comment]
@mikecowgill This may work but v3 is not 100% compatible with v2 which the plugin is currently designed to use. We haven't tested it either so you should definitely test this thoroughly in a safe environment before using it in a production environment.

As for the future of this issue, we're all for moving to v3 of the SDK but the main hold up is the bump in the required version of PHP to 5.5. Before we can make the switch we need to give a fair warning to allow people a chance to upgrade since it is a substantial jump.

The core issue here is essentially a conflict of simultaneously loading different versions of the same SDK which are competing for the same namespace. Even if we upgrade to v3, there will still be the possibility of conflict with other plugins that use v2, or different versions of v3. The solution that we're working on is to bring the SDK that we need under our namespace and basically get out of the way so you are free to use whatever version of the SDK you want for your own needs.

No ETA at this time, but we're working on it 🛠

**Comment by mikecowgill** [comment]
@aaemnnosttv Understood and you're absolutely right. I've been testing this change only in a local environment for a project with a dependency on the v3 sdk.
Initial tests are looking ok as all assets are being uploaded to s3, in their proper buckets and with wp urls updated.
I'm only testing the S3 upload functionality but have not yet had any issues.

**Comment by ianmjones** [comment]
This was resolved with WP Offload S3 Lite v1.3 which bundled and namespaced what is needed of the AWS PHP SDK v2 so that it no longer clashes with other instances of the AWS PHP SDK, v3 or otherwise.

See the release blog post for details: https://deliciousbrains.com/wp-offload-s3-1-6-released/

Note: We are also working on switching to using v3 of the AWS PHP SDK too and hope to have it in the next release.
